### ISSUE TITLE
[FEA] DataFrame level support for some of the critical functions such as value_counts and unique

### ISSUE THREAD

**Issue by oyilmaz-nvidia** [rapidsai/cudf#5048]
**Is your feature request related to a problem? Please describe.**
I wish cuDF could run some of the Series level function like value_counts, unique, mean, std, var, min, max, etc in the DataFrame level. 

As an example, we need a version of "value_counts" function in the DataFrame level that runs on the columns in parallel. Right now, "value_counts" (and other functions as well) function is only available in the Series level which only allows us to run this function one column at a time. For instance, if we have 'C1', 'C2', 'C3', ..., 'C20' as categorical columns, we have to call the value_counts for 'C1', then for 'C2', ...., then for 'C20' one by one. I think this sequential computing approach is causing us to lose significant amount of time because we can calculate the value counts for 'C1', 'C2', 'C3', ..., 'C20' simultaneously in one CUDA kernel.

Also, some of these operations can be possibly fused into one or couple of CUDA kernels.

**Describe the solution you'd like**
Creating CUDA kernels for these functions to run on columns of DataFrame in parallel.

**Describe alternatives you've considered**
One easy and fast solution might be to create N number of streams where N is the number of columns in the DataFrame. Then, N number of kernel calls can be made using these streams. This might not be as fast as the one kernel approach since there will be a delay to create streams but it can provide a level of speedup. Maybe streams can be created at the beginning of a program in the cudf level and these streams can be used whenever need.

**Comment by oyilmaz-nvidia** [comment]
@EvenOldridge @benfred ++ for visibility

**Comment by kkraus14** [comment]
cc @harrism since this needs to be basically entirely driven by libcudf

**Comment by harrism** [comment]
@oyilmaz-nvidia I'll have to look up what value_counts does (which I assume is a histogram), but other than unique, the others are all reductions, and so should be doable. However, in libcudf, we might end up just looping over the columns to do it, because of the way our implementation works. I suppose since Python loops are slow, that might still be better.

**Comment by oyilmaz-nvidia** [comment]
@harrism Thanks Mark for checking this.

1) Will create a separate issue for each function
2) Like you said, maybe you can loop over the columns with a stream per column. In that case, we might get a speedup.
3) For the histogram and unique on tables, we want the columns to be treated independently. 

Thanks!

**Comment by kkraus14** [comment]
> I'll have to look up what value_counts does (which I assume is a histogram)

It's actually a reduce by key where the keys are the values in the column and the reduction is a count (groupby with count aggregation).


> However, in libcudf, we might end up just looping over the columns to do it, because of the way our implementation works.

I may be mistaken but I believe the ask here is more along the lines of a fused kernel or some form of overlapping execution.

**Comment by oyilmaz-nvidia** [comment]
@kkraus14 Thanks for mentioning the fused kernel.

The way I see is that there are 3 different ways to implement these functions in the DataFrame level. Obviously, some of them will be easy in terms of development but might give us a limited speedup. 

1) Easiest and fastest way in terms of development time: loop over the columns and call the kernel with a separate stream. I guess we can do this for all the functions (min, max, std, var, value_counts, etc)

2) Create a kernel for a function that can run reduction (or whatever the operation needed) for all the columns in parallel. I think this should be faster than the number 1.

3) Create a kernel that can fuse these functions wherever possible in the DataFrame level. For instance, if we need std, var, and mean, I think we can calculate these functions in one fused kernel. I think this should be the fastest. 

Having number 3 would be super nice but I can see that will take some time. So, it's obviously your decision what to do next.

Thanks for considering this feature in cudf.

**Comment by kkraus14** [comment]
> Easiest and fastest way in terms of development time: loop over the columns and call the kernel with a separate stream. I guess we can do this for all the functions (min, max, std, var, value_counts, etc)

Creating streams is expensive and libcudf doesn't currently have a stream pool implemented as of now so this isn't entirely straightforward.

**Comment by harrism** [comment]
> > I'll have to look up what value_counts does (which I assume is a histogram)
> 
> It's actually a reduce by key where the keys are the values in the column and the reduction is a count (groupby with count aggregation).

Yes, that is a histogram (count of each unique value).

Keep in mind that `thrust::reduce_by_key` requires sorting the values, since it really only considers contiguous identical keys. That may or may not be faster than other approaches, depending on cardinality.

In general all 3 suggestions make assumptions about libcudf internals. Since the operations requested are of a few different classes, they are likely to have different implementations. 

* min/max -- simple reduction
* mean / std / var -- multistage reduction
* unique
* histogram

These 3 (or 4) each deserve separate consideration, so a separate issue would help.

**Comment by argenisleon** [comment]
Hi guys,

Any way to track down this? Were the separated issues created?

**Comment by harrism** [comment]
No, assigning to @oyilmaz-nvidia to create them.

**Comment by oyilmaz-nvidia** [comment]
Created https://github.com/rapidsai/cudf/issues/5169, https://github.com/rapidsai/cudf/issues/5170, https://github.com/rapidsai/cudf/issues/5171, https://github.com/rapidsai/cudf/issues/5172, https://github.com/rapidsai/cudf/issues/5173, https://github.com/rapidsai/cudf/issues/5174, https://github.com/rapidsai/cudf/issues/5175, https://github.com/rapidsai/cudf/issues/5176, https://github.com/rapidsai/cudf/issues/5177, https://github.com/rapidsai/cudf/issues/5178.

**Comment by devavret** [comment]
@oyilmaz-nvidia , Can you give some usage code we can profile. Would like to see how much of the performance loss is at the python level i.e. how much we could gain by simply looping over the series level reductions in libcudf.

**Comment by harrism** [comment]
> Created #5169, #5170, #5171, #5172, #5173, #5174, #5175, #5176, #5177, #5178.

Now I regret my request. 

> These 3 (or 4) each deserve separate consideration, so a separate issue would help.

Note not 10. Similar functions can be done together as explained above. The point is just that unique and value_counts() are very different from the other functions.

**Comment by oyilmaz-nvidia** [comment]
--- Now I regret my request.
:) Sorry.

--- Note not 10. Similar functions can be done together as explained above. The point is just that unique and value_counts() are very different from the other functions.

Missed that part. I thought you'd like get an issue for each function.

**Comment by GregoryKimball** [comment]
Closing this in favor of #5170, and in light of all the tasks already completed.
